### ISSUE TITLE
ec2_group no longer "prefers" names in the same VPC

### ISSUE THREAD

**Issue by bloob00k** [ansible/ansible#45782]
##### SUMMARY

If a rule definition specifies a group_name as the source, and groups with that name exist in multiple VPCs, the chosen group appears to be "random"; previously the group with that name in the VPC where the new group is being created was used.  Now it is picking up a group in a different VPC, causing the group creation to fail with "The error was: ClientError: An error occurred (InvalidGroup.NotFound) when calling the AuthorizeSecurityGroupIngress operation: You have specified two resources that belong to different networks."

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

ec2_group

##### ANSIBLE VERSION

```paste below
ansible 2.6.3
  config file = /home/pfletcher/.ansible.cfg
  configured module search path = [u'/home/pfletcher/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Aug  4 2017, 00:39:18) [GCC 4.8.5 20150623 (Red Hat 4.8.5-16)]
```

##### CONFIGURATION

```paste below
[DEPRECATION WARNING]: DEFAULT_SUDO_FLAGS option, In favor of Ansible Become, which is a generic framework. See become_flags. , use become instead. This
feature will be removed in version 2.8. Deprecation warnings can be disabled by setting deprecation_warnings=False in ansible.cfg.
ANSIBLE_PIPELINING(/home/pfletcher/.ansible.cfg) = True
ANSIBLE_SSH_ARGS(/home/pfletcher/.ansible.cfg) = -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o ControlMaster=auto -o ControlPersist=30m
ANSIBLE_SSH_CONTROL_PATH(/home/pfletcher/.ansible.cfg) = %(directory)s/%%h-%%r
DEFAULT_CALLBACK_WHITELIST(/home/pfletcher/.ansible.cfg) = ['profile_tasks']
DEFAULT_REMOTE_USER(/home/pfletcher/.ansible.cfg) = pfletcher
DEFAULT_SUDO_FLAGS(/home/pfletcher/.ansible.cfg) = -H -S -i
DEFAULT_TIMEOUT(/home/pfletcher/.ansible.cfg) = 20
```

##### OS / ENVIRONMENT

Host OS: CentOS Linux release 7.2.1511 (Core)

##### STEPS TO REPRODUCE

Two VPCs in the same account.  Both have a security group called existing-group-1.  First VPC also has a group called existing-group-2.  (Note: if I remove the existing-group-2 rule below this playbook succeeds, but that may just be chance, not sure).

Run the playbook below.  vpc_id is the ID of ther first VPC - the one which has both of the existing groups in it.

```yaml
- hosts: 127.0.0.1
  tasks:
    - ec2_group:
        name:            Test
        description:     Test group
        region:          us-east-1
        vpc_id:          vpc-xxxxxxx
        rules:
          - proto:        all
            group_name:   existing-group-2
          - proto:        udp
            from_port:    88
            to_port:      88
            group_name:   existing-group-1

ansible-playbook group-bug.yml -vvvv
```

##### EXPECTED RESULTS

The group should be successfully created, with rules referencing the two existing security groups in the specified VPC.

##### ACTUAL RESULTS

Playbook fails as below.

```paste below
The full traceback is:
Traceback (most recent call last):
  File "/tmp/ansible_oj8wTi/ansible_module_ec2_group.py", line 700, in authorize
    client.authorize_security_group_ingress(GroupId=group_id, IpPermissions=ip_permissions)
  File "/usr/lib/python2.7/site-packages/botocore/client.py", line 314, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/usr/lib/python2.7/site-packages/botocore/client.py", line 612, in _make_api_call
    raise error_class(parsed_response, operation_name)
ClientError: An error occurred (InvalidGroup.NotFound) when calling the AuthorizeSecurityGroupIngress operation: You have specified two resources that belong to different networks.

fatal: [127.0.0.1]: FAILED! => {
    "boto3_version": "1.7.54",
    "botocore_version": "1.10.54",
    "changed": false,
    "error": {
        "code": "InvalidGroup.NotFound",
        "message": "You have specified two resources that belong to different networks."
    },
    "invocation": {
        "module_args": {
            "aws_access_key": null,
            "aws_secret_key": null,
            "description": "Test group",
            "ec2_url": null,
            "group_id": null,
            "name": "Test",
            "profile": null,
            "purge_rules": true,
            "purge_rules_egress": true,
            "purge_tags": true,
            "region": "us-east-1",
            "rules": [
                {
                    "group_name": "existing-group-2",
                    "proto": "all"
                },
                {
                    "from_port": 88,
                    "group_name": "existing-group-1",
                    "proto": "udp",
                    "to_port": 88
                }
            ],
            "rules_egress": null,
            "security_token": null,
            "state": "present",
            "tags": null,
            "validate_certs": true,
            "vpc_id": "vpc-xxxxxxx"
        }
    },
    "msg": "Unable to authorize ingress rules: [{'ToPort': 88, 'IpProtocol': u'udp', 'UserIdGroupPairs': [{'GroupId': 'sg-f4ca2ebf'}], 'FromPort': 88}, {'IpProtocol': u'-1', 'UserIdGroupPairs': [{'GroupId': 'sg-0db76768a3d9b253a'}]}]: An error occurred (InvalidGroup.NotFound) when calling the AuthorizeSecurityGroupIngress operation: You have specified two resources that belong to different networks.",
    "response_metadata": {
        "http_headers": {
            "connection": "close",
            "date": "Tue, 18 Sep 2018 12:11:07 GMT",
            "server": "AmazonEC2",
            "transfer-encoding": "chunked"
        },
        "http_status_code": 400,
        "request_id": "29f19b7b-b3aa-4604-aad5-ced1f99787b7",
        "retry_attempts": 0
    }
}
```

**Comment by bloob00k** [comment]
PS rolling back to commit from April 24th (prior to refactor) fixes the issue.

**Comment by ansibot** [comment]
cc @adq
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by s-hertel** [comment]
Thanks for opening the issue @bloob00k. I was able to reproduce this behavior and submitted #45787 to fix it. If you can test double-check that it fixes the issue for you I would appreciate it a lot. :-)

**Comment by bloob00k** [comment]
Hi there,

Thanks for the fast turn-around!  I have tried to verify the fix but am
unable to do so due to environmental issues.  My environment is set up for
2.6.3, and everything is installed with yum / RPMs.  This version of the
module requires a newer version of boto3 so far as I can tell (wants
boto3_error_code).  I tried to update that with pip - despite myself - but
it wasn't enough, and by then I had already messed up my environment quite
badly, so I decided to quit while I still could.

If it gets back-ported into 2.6 I'd be glad to have another go.

Paul
